Data Structures for Parser

mattgathu opened the pull request:
Add some data structures need in implementing the deterministic parser.

1. Trie
--
A non-generic trie implementation for efficient storage of utterances
and their mapped outputs.
The implementation is influenced by the `trie` and `radix-trie` crates:
- [trie](https://github.com/contain-rs/trie)
- [radix trie](https://github.com/michaelsproul/rust_radix_trie)

2. Symbol Table
--
A non-generic table for storing string symbols, this basically
a wrapper around Rust's stdlib's HashMap with additional hashing
for generating symbol keys.

There is also a corresponding Foreign Function Interface to expose
the data structures apis to the C FFI.

adrienball commented:
@mattgathu,
I have not thoroughly reviewed your PR yet, but I've seen that you have included the FFI into the main lib. I think we should rather extract the FFI into a separate crate to avoid including unnecessary dependencies. You can look at what has been done in [snips-nlu-parsers](https://github.com/snipsco/snips-nlu-parsers) or [snips-nlu-ontology](https://github.com/snipsco/snips-nlu-ontology).

On another topic, don't you think that we should simplify a bit the FFI API and handle the symbol table internally ? This was my first reaction when I saw all the FFI functions.
What do you think?

Cheers

mattgathu commented:
@adrienball yeah I could move the ffi to a separate crate. Initially the ffi layer was very minimal and warrant an external crate but it would be consistent to have it externally.

I wasn't sure about internally coupling the **Trie** with the **SymbolTable** since there are conceptually independent of each other. If you think this is a good idea, we can definitely do that and only exposed the Trie ffi interface only.

adrienball commented:
We can keep *Trie* and *SymbolTable* uncoupled, and create a parent structure encapsulating them, and providing a simple top level API.

adrienball commented on ffi/src/lib.rs:
```diff
@@ -0,0 +1,85 @@
+pub unsafe extern "C" fn trie_map_insert(
+    map_ptr: *mut StringTrieMap,
+    k: *const i8,
+    v: *const i8,
```
Shouldn't it be:
```rust
k: *const libc::c_char,
v: *const libc::c_char,
```
?

adrienball commented on ffi/src/lib.rs:
```diff
@@ -0,0 +1,85 @@
+/// load trie map from the file system
+pub extern "C" fn trie_map_load(
+    map_ptr: *mut *const StringTrieMap,
+    path: *const i8,
```
Same here regarding `*const libc::c_char` instead of `*const i8`

adrienball commented on src/string_trie_map.rs:
```diff
@@ -0,0 +1,104 @@
+            } 
+        }
+        self.trie.remove(key);
+    }
```
This function could return an `Option<String>` containing the corresponding removed value, if any.

adrienball commented on src/string_trie_map.rs:
```diff
@@ -0,0 +1,104 @@
+        self.table.dump(table_path)?;
+        self.trie.dump(trie_path)?;
+        Ok(())
+    }
```
I'm wondering if we should not create a directory containing the two files instead. Otherwise the path to provide when calling `load` may be a bit unclear, WDYT?

mattgathu commented:
Consider using string as trie keys - why not

mattgathu commented:
Why is the trie implementation hashing the keys? We can avoid that

mattgathu commented on src/trie.rs:
```diff
@@ -0,0 +1,409 @@
+    }
+}
+
+impl From<&[u8]> for Key {
```
refactor key not to have hashing.

mattgathu commented:
@adrienball I have addressed the issues with the review:
- dependencies version
- loading / dumping from directory
- trie key bug

You can have another look 🙂

mattgathu commented:
Closing this after being superseded by https://github.com/snipsco/snips-nlu-rs/pull/121
